### ISSUE TITLE
Issues on windows (conda smithy regenerate)

### ISSUE THREAD

**Issue by janschulz** [conda-forge/conda-smithy#350]
I tried to fix the problems in https://travis-ci.org/conda-forge/pypandoc-feedstock/jobs/171846025 by rerendering the [pypandoc-feedstock](https://github.com/conda-forge/pypandoc-feedstock), but I ran into two issues (I'm on windows 7):

```
λ conda smithy regenerate
Traceback (most recent call last):
  File "C:\portabel\miniconda\Scripts\conda-smithy-script.py", line 5, in <module>
    sys.exit(conda_smithy.cli.main())
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\cli.py", line 289, in main
    args.subcommand_func(args)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\cli.py", line 200, in __call__
    configure_feedstock.main(args.feedstock_directory)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\configure_feedstock.py", line 455, in main
    copy_feedstock_content(forge_dir)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\configure_feedstock.py", line 361, in copy_feedstock_content
    'README'
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\configure_feedstock.py", line 352, in copytree
    copy_file(s, d)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\feedstock_io.py", line 74, in copy_file
    repo.index.add([dst])
  File "C:\portabel\miniconda\lib\site-packages\git\index\base.py", line 745, in add
    entries_added.extend(self._entries_for_paths(paths, path_rewriter, fprogress, entries))
  File "C:\portabel\miniconda\lib\site-packages\git\util.py", line 55, in wrapper
    return func(self, *args, **kwargs)
  File "C:\portabel\miniconda\lib\site-packages\git\index\util.py", line 84, in set_git_working_dir
    return func(self, *args, **kwargs)
  File "C:\portabel\miniconda\lib\site-packages\git\index\base.py", line 631, in _entries_for_paths
    entries_added.append(self._store_path(filepath, fprogress))
  File "C:\portabel\miniconda\lib\site-packages\git\index\base.py", line 599, in _store_path
    istream = self.repo.odb.store(IStream(Blob.type, st.st_size, stream))
  File "C:\portabel\miniconda\lib\site-packages\gitdb\db\loose.py", line 230, in store
    remove(obj_path)
PermissionError: [WinError 5] Zugriff verweigert: 'c:\\data\\external\\conda-packages\\pypandoc-feedstock\\.git\\objects\\3d\\c1e2f50daaa760feec968c7501f95973f33082'
```

["Zugriff verweigert" == "access denied"]

This problem was reproducible but went away after doing a `git gc`.

Then I ran into this:

```
λ conda smithy regenerate
Fetching package metadata ...............
Traceback (most recent call last):
  File "C:\portabel\miniconda\Scripts\conda-smithy-script.py", line 5, in <module>
    sys.exit(conda_smithy.cli.main())
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\cli.py", line 289, in main
    args.subcommand_func(args)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\cli.py", line 200, in __call__
    configure_feedstock.main(args.feedstock_directory)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\configure_feedstock.py", line 457, in main
    render_run_docker_build(env, config, forge_dir)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\configure_feedstock.py", line 110, in render_run_docker_build
    mode = get_mode_file(each_target_fname)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\feedstock_io.py", line 32, in get_mode_file
    blob = get_file_blob(repo, filename)
  File "C:\portabel\miniconda\lib\site-packages\conda_smithy\feedstock_io.py", line 25, in get_file_blob
    blob = next(idx.iter_blobs(lambda _: _[1].path == rel_filepath))[1]
StopIteration
```

No idea what to do here :-(

```
λ conda smithy --version
1.4.3
```

**Comment by MSeifert04** [comment]
I had the same problem with the Access denied and I fixed it by removing the readonly attribute (schreibgeschützt) for the `xxx-feedstock/.git` folder and subfolder.

No idea about the second one.

**Comment by jakirkham** [comment]
> I fixed it by removing the readonly attribute (schreibgeschützt) for the `xxx-feedstock/.git` folder and subfolder.

Interesting. Is this a regular thing on Windows to have readonly attributes on `.git` @janschulz @MSeifert04 ?

**Comment by jakirkham** [comment]
That second error is interesting though, @janschulz . Basically that means we have tried to ask `git` about permission info on a file that doesn't exist. I'm not entirely sure how that could happen. I have tried to reproduce it on my Mac, but seem unable to. Maybe there is something subtle about timing of the IO operations in your configuration. Could you please tell me what `git status` says after it fails? Also more information about how you are running on Windows would be very helpful.

TBH I'm amazed either of you are either to re-render on Windows given issue ( https://github.com/conda-forge/conda-smithy/issues/281 ) has been a blocker for everyone thus far. Is there a workaround that you could share with them?

**Comment by MSeifert04** [comment]
@jakirkham I don't know if I just did something wrong when setting up `git` (or Windows :smile:) but the files in `.git/objects/...` all have the "readonly" attribute (all of my projects!).

I haven't been able to work around the mentioned issue (#281), so I resurrected my VirtualBox and did the re-rendering there and hit another Bug (https://github.com/conda-forge/iteration_utilities-feedstock/pull/1#discussion_r85696973).

**Comment by jakirkham** [comment]
Sure, the files inside `.git/objects/...` are readonly for me too. Though that is as expected as those should never change. I thought you were saying the entirety of `.git/` was readonly, which would definitely not be good.

xref: https://git-scm.com/book/en/v2/Git-Internals-Git-Objects

**Comment by MSeifert04** [comment]
@jakirkham But it looks like it is trying to remove a file there: `c:\\data\\external\\conda-packages\\pypandoc-feedstock\\.git\\objects\\3d\\c1e2f50daaa760feec968c7501f95973f33082'`

```
File "C:\portabel\miniconda\lib\site-packages\gitdb\db\loose.py", line 230, in store
    remove(obj_path)
PermissionError: [WinError 5] Zugriff verweigert: 'c:\\data\\external\\conda-packages\\pypandoc-feedstock\\.git\\objects\\3d\\c1e2f50daaa760feec968c7501f95973f33082'
```

**Comment by jakirkham** [comment]
Sure, so if a file is readonly one can still delete it. They just can't write to it. For instance, if one tried to remove one of these files via `rm` on a POSIX OS, they would have to change it to `rm -f` for the removal to proceed. However, if I try the same option in Python (e.g. `os.remove` as here), the removal proceeds without issues. So it is unclear to me why that doesn't happen on Windows. Again knowing more about your ( @MSeifert04 ) and @janschulz 's Windows configurations would be very helpful.

**Comment by jakirkham** [comment]
Should add we are using a very old copy of GitPython with `conda-smithy`. ( https://github.com/conda-forge/conda-smithy/issues/298 ) TBH I don't know why that is the case. We are also using a very old copy of GitDB. It is possible these issue are fixed by upgrading. Would be good if @pelson could weigh in on these issues, but I don't think that is going to happen soon.

**Comment by MSeifert04** [comment]
@jakirkham But OSError 5 means that it's something about permissions. 

Maybe it doesn't use `\F` for forced-delete on Windows?

But I'm just guessing, I don't have that much experience with git/windows to really know how to debug this issue or even know what could be important.

**Comment by jakirkham** [comment]
> But OSError 5 means that it's something about permissions. 

Noted. Definitely something is not right. It is just going to take a bit more work to narrow down where.

> Maybe it doesn't use `\F` for forced-delete on Windows?

Again we are getting into the implementation details of `os.remove` in Python on Windows. So this would require going through the API docs (and potentially code) with a fine tooth comb.

> But I'm just guessing, I don't have that much experience with git/windows to really know how to debug this issue or even know what could be important.

Actually you still can help. The fact that you are actually using Windows is already infinitely more help than you think. 😉

If you could try patching the installed `gitdb` as mentioned in the [comment above]( https://github.com/conda-forge/conda-smithy/issues/350#issuecomment-257328169 ), that would really help me understand whether this issue will be fixed by upgrading `gitdb` or not.

**Comment by MSeifert04** [comment]
@jakirkham The problem is after I changed the permissions I can't reproduce the `OSError` anymore. Even deleting and re-cloning it didn't bring it back.

**Comment by jakirkham** [comment]
Ok, well, if you are able to reproduce it again, please give that a try. Should be able to uninstall and reinstall `gitdb` to clear any patch changes as needed.

**Comment by jakirkham** [comment]
I wonder if a less aggressive solution other that `git gc` would work. Perhaps `git update-index -q --refresh` would do the trick. Not sure if that would run into other issues too, but that might be benign enough that we could just run it before each re-rendering to ensure we don't run into the first issue.

**Comment by MSeifert04** [comment]
@jakirkham How can I uninstall `gitdb`, that also wants to uninstall `conda-smithy` if I use `conda remove gitdb` (even after I pip-installed the 2.0 gitdb version).

**Comment by jakirkham** [comment]
Right, so I think you will need to manually edit ` gitdb/db/loose.py` based on commit ( https://github.com/gitpython-developers/gitdb/commit/2389b75280efb1a63e6ea578eae7f897fd4beb1b ).

Sorry it is kind of a pain. I'm just very unclear about `gitpython`/`gitdb` compatibility ATM. So don't want to mess with `gitdb` more generally until we understand that better.

**Comment by MSeifert04** [comment]
ok, if I encounter the issue again I'll do that. 👍

**Comment by jakirkham** [comment]
Not sure if this solves your issue or a related issue @janschulz , but PR ( https://github.com/conda-forge/conda-smithy/pull/359 ) solves an issue that encounters the same exception.

**Comment by jakirkham** [comment]
I think we may have fixed item 2 mentioned by the OP.

However, item 1 is still technically outstanding. Though we are currently unable to reproduce it. If we can reproduce item 1, will happily reopen and attempt to address it.

**Comment by jakirkham** [comment]
Also have deployed fixes for various things in `1.4.6`. Please give it a try. If this issue crops back up, we can certainly reopen and discuss failing cases.
